go-controller/pkg/ovn/egressgw.go

```diff
@@ -420,15 +405,25 @@ func (oc *DefaultNetworkController) deletePodGWRoutesForNamespace(pod *kapi.Pod,
 // are given, all routes for the namespace are deleted.
 func (oc *DefaultNetworkController) deleteGWRoutesForNamespace(namespace string, matchGWs sets.Set[string]) error {
 	deleteAll := (matchGWs == nil || matchGWs.Len() == 0)
+
+	policyGWIPs, err := oc.apbExternalRouteController.GetDynamicGatewayIPsForTargetNamespace(namespace)
+	if err != nil {
+		return err
+	}
+	policyStaticGWIPs, err := oc.apbExternalRouteController.GetStaticGatewayIPsForTargetNamespace(namespace)
+	if err != nil {
+		return err
+	}
+	policyGWIPs.Union(policyStaticGWIPs)
 	for _, routeInfo := range oc.getRouteInfosForNamespace(namespace) {
 		routeInfo.Lock()
-		if routeInfo.deleted {
+		if routeInfo.Deleted {
 			routeInfo.Unlock()
 			continue
 		}
-		for podIP, routes := range routeInfo.podExternalRoutes {
+		for podIP, routes := range routeInfo.PodExternalRoutes {
 			for gw, gr := range routes {
-				if deleteAll || matchGWs.Has(gw) {
+				if (deleteAll || matchGWs.Has(gw)) && !policyGWIPs.Has(gw) {
```

Comment:
cool. I guess hypothetically there could be a race here where GetDynamicGatewayIPsForTargetNamespace gives you the IP addresses at one point in time, and while you go to deletePodGWRoute, it is added by the other controller, and we accidentally remove it here? 

Not sure if this is a case we care about if we don't plan on full support for the legacy annotation with simultaneous use of the CRD. Perhaps we just give a proper upgrade method from annotations to CRD to avoid this potential issue.

To properly fix the simultaneous use we would probably have to lock the route policy controller for the affected keys while we do this.